### PR TITLE
[datadog_service_level_objective] Add support for time-slice SLOs

### PR DESCRIPTION
**Pull request by ddjamesfrullo**
Overview
----------------
This adds support for time-slice SLOs to the Terraform provider. Here's what the terraform resource will look like:
```
resource "datadog_service_level_objective" "time_slice_slo_1" {
  name              = "Terraform test for Time Slice"
  type              = "time_slice"
  description       = "Created as part of a manual test of Terrform provider changes to support time-slice SLO"
  sli_specification {
    time_slice {
      query {
        formula {
          formula_expression = "(query1-query2)/query1"
        }
        query {
          metric_query {
            name        = "query1"
            query       = "sum:trace.grpc.server.hits{service:monitor-history-reader}"
          }
        }
        query {
          metric_query {
            name        = "query2"
            query       = "sum:trace.grpc.server.errors{service:monitor-history-reader}"
          }
        }
      }
      comparator = ">"
      threshold  = 0.99
    }
  }

  thresholds {
    timeframe = "7d"
    target    = 99.0
    warning   = 99.5
  }

  timeframe         = "7d"
  target_threshold  = 99.0
  warning_threshold = 99.5

  tags              = ["test:true"]
}
```

Testing
---------------
I added a new acceptance test for time-slice SLO creation in this review, but I also manually tested a number of cases to make sure I could create SLOs in org 2, and also to see what validation errors look like. I specifically added new validation that happens during apply to make sure required fields aren't missing. Here's an example of those errors:
```
╷
│ Error: sli_specification is required for time slice SLOs
│ 
│   with datadog_service_level_objective.time_slice_slo_validation_error_6,
│   on slo.tf line 204, in resource "datadog_service_level_objective" "time_slice_slo_validation_error_6":
│  204: resource "datadog_service_level_objective" "time_slice_slo_validation_error_6" {
│ 
╵
╷
│ Error: query is required for metric SLOs
│ 
│   with datadog_service_level_objective.time_slice_slo_validation_error_7,
│   on slo.tf line 225, in resource "datadog_service_level_objective" "time_slice_slo_validation_error_7":
│  225: resource "datadog_service_level_objective" "time_slice_slo_validation_error_7" {
│ 
╵
╷
│ Error: monitor_ids is required for monitor SLOs
│ 
│   with datadog_service_level_objective.time_slice_slo_validation_error_8,
│   on slo.tf line 242, in resource "datadog_service_level_objective" "time_slice_slo_validation_error_8":
│  242: resource "datadog_service_level_objective" "time_slice_slo_validation_error_8" {
│ 
╵
```

### CODE REVIEW

**Comment by bogaertg** [comment]
👍

File: datadog/resource_datadog_service_level_objective.go [hunk 1]
```diff
@@ -134,7 +189,7 @@ func resourceDatadogServiceLevelObjective() *schema.Resource {
 					Type:          schema.TypeList,
 					MaxItems:      1,
 					Optional:      true,
-					ConflictsWith: []string{"monitor_ids", "groups"},
+					ConflictsWith: []string{"monitor_ids", "sli_specification", "groups"},
```
Comment (ddjamesfrullo) [inline datadog/resource_datadog_service_level_objective.go:192]: Prevents users from specifying `query` for time slice SLOs.

File: datadog/resource_datadog_service_level_objective.go [hunk 2]
```diff
@@ -160,10 +215,48 @@ func resourceDatadogServiceLevelObjective() *schema.Resource {
 				"monitor_ids": {
 					Type:          schema.TypeSet,
 					Optional:      true,
-					ConflictsWith: []string{"query"},
+					ConflictsWith: []string{"query", "sli_specification"},
```
Comment (ddjamesfrullo) [inline datadog/resource_datadog_service_level_objective.go:218]: Prevents users from specifying `sli_specificaton` for monitor SLOs.

File: datadog/resource_datadog_service_level_objective.go [hunk 3]
```diff
@@ -160,10 +215,48 @@ func resourceDatadogServiceLevelObjective() *schema.Resource {
...
 					Elem:          &schema.Schema{Type: schema.TypeInt, MinItems: 1},
 				},
+
+				// Time-Slice SLO
+				"sli_specification": {
+					Type:          schema.TypeList,
+					MinItems:      1,
+					MaxItems:      1,
+					Optional:      true,
+					ConflictsWith: []string{"query", "monitor_ids", "groups"},
```
Comment (ddjamesfrullo) [inline datadog/resource_datadog_service_level_objective.go:229]: Cannot specify `query`, `monitor_ids` or `groups` for time-slice SLOs.

File: docs/resources/service_level_objective.md [hunk 4]
```diff
@@ -124,6 +158,48 @@ Required:
 - `denominator` (String) The sum of the `total` events.
 - `numerator` (String) The sum of all the `good` events.
 
+
+<a id="nestedblock--sli_specification"></a>
+### Nested Schema for `sli_specification`
+
+Required:
+
+- `time_slice` (Block List, Min: 1, Max: 1) The time slice condition, composed of 3 parts: 1. the timeseries query, 2. the comparator, and 3. the threshold. (see [below for nested schema](#nestedblock--sli_specification--time_slice))
```
Comment (aliciascott) [inline docs/resources/service_level_objective.md:167]:
```suggestion
- `time_slice` (Block List, Min: 1, Max: 1) The time slice condition, composed of 3 parts: 1. The timeseries query, 2. The comparator, and 3. The threshold. (see [below for nested schema](#nestedblock--sli_specification--time_slice))
```

File: docs/resources/service_level_objective.md [hunk 5]
```diff
@@ -124,6 +158,48 @@ Required:
...
+
+<a id="nestedblock--sli_specification--time_slice"></a>
+### Nested Schema for `sli_specification.time_slice`
+
+Required:
+
+- `comparator` (String) The comparator used to compare the SLI value to the threshold Valid values are `>`, `>=`, `<`, `<=`.
```
Comment (aliciascott) [inline docs/resources/service_level_objective.md:174]:
```suggestion
- `comparator` (String) The comparator used to compare the SLI value to the threshold. Valid values are `>`, `>=`, `<`, `<=`.
```

File: docs/resources/service_level_objective.md [hunk 6]
```diff
@@ -124,6 +158,48 @@ Required:
...
+
+- `formulas` (Block List, Min: 1, Max: 1) A list that contains exactly one formula, as only a single formula may be used in a time-slice SLO (see [below for nested schema](#nestedblock--sli_specification--time_slice--query--formulas))
+- `queries` (Block List, Min: 1) A list of queries that are used to calculate the SLI value (see [below for nested schema](#nestedblock--sli_specification--time_slice--query--queries))
+
+<a id="nestedblock--sli_specification--time_slice--query--formulas"></a>
+### Nested Schema for `sli_specification.time_slice.query.formulas`
+
+Required:
+
+- `formula` (String) The formula string, which is an expression involving named queries
```
Comment (aliciascott) [inline docs/resources/service_level_objective.md:191]:
```suggestion
- `formula` (String) The formula string, which is an expression involving named queries.
```

File: docs/resources/service_level_objective.md [hunk 7]
```diff
@@ -124,6 +158,48 @@ Required:
...
+
+
+<a id="nestedblock--sli_specification--time_slice--query--queries"></a>
+### Nested Schema for `sli_specification.time_slice.query.queries`
+
+Required:
+
+- `data_source` (String) The data source to use for the query, which must be `metrics` Valid values are `metrics`.
```
Comment (aliciascott) [inline docs/resources/service_level_objective.md:199]:
```suggestion
- `data_source` (String) The data source to use for the query, which must be `metrics`. Valid values are `metrics`.
```

**Review by aliciascott — APPROVED** [review]
Just some small punctuation and capitalization suggestions, otherwise approved!

**Review by mmeyer724 — APPROVED** [review]

File: datadog/resource_datadog_service_level_objective.go [hunk 8]
```diff
@@ -216,7 +322,42 @@ func resourceDatadogServiceLevelObjectiveCustomizeDiff(ctx context.Context, diff
...
+						datadogV1.FormulaAndFunctionMetricQueryDefinitionAsSLODataSourceQueryDefinition(
+							datadogV1.NewFormulaAndFunctionMetricQueryDefinition(*dataSource, name, query)))
+				}
+			}
+		}
+	}
+	return ret
+}
+
+func buildServiceLevelObjectiveStructs(d *schema.ResourceData) (*datadogV1.ServiceLevelObjective, *datadogV1.ServiceLevelObjectiveRequest, error) {
```
Comment (ddjamesfrullo) [inline datadog/resource_datadog_service_level_objective.go:360]: Add the ability to return errors.

File: datadog/resource_datadog_service_level_objective.go [hunk 9]
```diff
@@ -242,6 +383,8 @@ func buildServiceLevelObjectiveStructs(d *schema.ResourceData) (*datadogV1.Servi
 			}
 			slo.SetMonitorIds(s)
 			slor.SetMonitorIds(s)
+		} else {
+			return nil, nil, fmt.Errorf("monitor_ids is required for monitor SLOs")
```
Comment (ddjamesfrullo) [inline datadog/resource_datadog_service_level_objective.go:387]: New checks that ensure useful error messages when logically required fields are missing.

**Review by nkzou — APPROVED** [review]

**Pull request merged**